File: eval-score/node/src/assertion-checker.ts
import * as fs from 'fs';
import * as path from 'path';
import { EvalRow, Assertion, AssertionResult } from './types';

/**
 * EvalGen sidecar JSON format
 */
interface EvalGenSidecar {
  version: string;
  items: Array<{
    prompt: string;
    assertions?: Assertion[];
  }>;
}

/**
 * Load assertions from an EvalGen sidecar JSON file and attach them to eval rows.
 * Matches by prompt text (case-insensitive, trimmed).
 */
export function loadAssertionsFromSidecar(
  rows: EvalRow[],
  sidecarPath: string,
): EvalRow[] {
  const absPath = path.resolve(sidecarPath);
  if (!fs.existsSync(absPath)) {
    throw new Error(`Sidecar file not found: ${absPath}`);
  }

  const content = fs.readFileSync(absPath, 'utf-8');
  const sidecar: EvalGenSidecar = JSON.parse(content);

  if (!sidecar.items || !Array.isArray(sidecar.items)) {
    throw new Error('Invalid sidecar format: missing "items" array');
  }

  // Build a lookup map by normalized prompt
  const assertionMap = new Map<string, Assertion[]>();
  for (const item of sidecar.items) {
    if (item.prompt && item.assertions && item.assertions.length > 0) {
      const key = item.prompt.trim().toLowerCase();
      assertionMap.set(key, item.assertions);
    }
  }

  // Attach assertions to matching rows
  for (const row of rows) {
    const key = row.prompt.trim().toLowerCase();
    const assertions = assertionMap.get(key);
    if (assertions) {
      row.assertions = assertions;
    }
  }

  return rows;
}

/**
 * Evaluate a single assertion against an actual answer.
 */
export function evaluateAssertion(assertion: Assertion, actualAnswer: string): AssertionResult {
  const answer = actualAnswer.toLowerCase();

  switch (assertion.type) {
    case 'must_contain': {
      const target = assertion.value.toLowerCase();
      let passed: boolean;
      if (assertion.wholeWord) {
        // Word-boundary matching to avoid false positives on short values
        const regex = new RegExp(`\\b${target.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')}\\b`, 'i');
        passed = regex.test(actualAnswer);
      } else {
        passed = answer.includes(target);
      }
      return {
        assertion,
        passed,
        detail: passed
          ? `✅ Found "${assertion.value}"`
          : `❌ Missing "${assertion.value}"`,
      };
    }

    case 'must_contain_any': {
      const found = assertion.values.find(v => answer.includes(v.toLowerCase()));
      const passed = !!found;
      return {
        assertion,
        passed,
        detail: passed
          ? `✅ Found "${found}"`
          : `❌ None found: ${assertion.values.map(v => `"${v}"`).join(', ')}`,
      };
    }

    case 'must_not_contain': {
      const target = assertion.value.toLowerCase();
      const passed = !answer.includes(target);
      return {
        assertion,
        passed,
        detail: passed
          ? `✅ Correctly absent: "${assertion.value}"`
          : `❌ Unexpectedly found "${assertion.value}"`,
      };
    }

    default:
      return {
        assertion,
        passed: false,
        detail: `⚠️ Unknown assertion type`,
      };
  }
}

/**
 * Evaluate all assertions for a single eval row.
 */
export function evaluateRowAssertions(row: EvalRow): AssertionResult[] {
  if (!row.assertions || row.assertions.length === 0) return [];
  if (!row.actualAnswer || row.actualAnswer.startsWith('[ERROR:')) return [];

  return row.assertions.map(a => evaluateAssertion(a, row.actualAnswer));
}

/**
 * Evaluate assertions for all rows and attach results.
 */
export function evaluateAllAssertions(rows: EvalRow[]): EvalRow[] {
  for (const row of rows) {
    row.assertionResults = evaluateRowAssertions(row);
  }
  return rows;
}

File: eval-ui-winui3/src/EvalToolkit.EvalScore/Assertions/AssertionChecker.cs
using System.Text.RegularExpressions;
using EvalToolkit.EvalScore.Models;

namespace EvalToolkit.EvalScore.Assertions;

/// <summary>
/// Evaluates EvalGen-style assertions against a row's actual answer.
/// Mirrors <c>eval-score/node/src/assertion-checker.ts</c>.
///
/// <para>TS contract:
/// <list type="bullet">
///   <item><c>must_contain</c>: case-insensitive substring; with
///     <c>wholeWord=true</c>, word-boundary regex with all regex
///     metacharacters in the value escaped.</item>
///   <item><c>must_contain_any</c>: case-insensitive substring of ANY
///     listed value; details report the first match.</item>
///   <item><c>must_not_contain</c>: case-insensitive substring; passes
///     when absent.</item>
///   <item>Detail strings use the original-cased value, but match
///     against a lowercased haystack.</item>
///   <item>Rows with no actual answer or an <c>[ERROR:</c> prefix
///     produce an empty assertion-result list (not failures).</item>
/// </list></para>
/// </summary>
public static class AssertionChecker
{
    /// <summary>
    /// Evaluate a single assertion against the row's actual answer text.
    /// </summary>
    public static AssertionResult EvaluateAssertion(Assertion assertion, string actualAnswer)
    {
        ArgumentNullException.ThrowIfNull(assertion);
        actualAnswer ??= string.Empty;
        string lower = actualAnswer.ToLowerInvariant();

        switch (assertion.Type)
        {
            case AssertionType.MustContain:
                return EvaluateMustContain(assertion, actualAnswer, lower);
            case AssertionType.MustContainAny:
                return EvaluateMustContainAny(assertion, lower);
            case AssertionType.MustNotContain:
                return EvaluateMustNotContain(assertion, lower);
            default:
                return new AssertionResult
                {
                    Assertion = assertion,
                    Passed = false,
                    Detail = "⚠️ Unknown assertion type",
                };
        }
    }

    private static AssertionResult EvaluateMustContain(Assertion assertion, string actualAnswer, string lowerAnswer)
    {
        string value = assertion.Value ?? string.Empty;
        string target = value.ToLowerInvariant();
        bool passed;
        if (assertion.WholeWord == true)
        {
            // TS: new RegExp(`\\b${target.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')}\\b`, 'i')
            string escaped = Regex.Escape(target);
            var rx = new Regex($@"\b{escaped}\b", RegexOptions.IgnoreCase | RegexOptions.CultureInvariant);
            passed = rx.IsMatch(actualAnswer);
        }
        else
        {
            passed = lowerAnswer.Contains(target, StringComparison.Ordinal);
        }
        return new AssertionResult
        {
            Assertion = assertion,
            Passed = passed,
            Detail = passed
                ? $"✅ Found \"{value}\""
                : $"❌ Missing \"{value}\"",
        };
    }

    private static AssertionResult EvaluateMustContainAny(Assertion assertion, string lowerAnswer)
    {
        IReadOnlyList<string> values = assertion.Values ?? Array.Empty<string>();
        string? found = null;
        foreach (string v in values)
        {
            if (lowerAnswer.Contains(v.ToLowerInvariant(), StringComparison.Ordinal))
            {
                found = v;
                break;
            }
        }
        bool passed = found is not null;
        return new AssertionResult
        {
            Assertion = assertion,
            Passed = passed,
            Detail = passed
                ? $"✅ Found \"{found}\""
                : $"❌ None found: {string.Join(", ", values.Select(v => $"\"{v}\""))}",
        };
    }

    private static AssertionResult EvaluateMustNotContain(Assertion assertion, string lowerAnswer)
    {
        string value = assertion.Value ?? string.Empty;
        string target = value.ToLowerInvariant();
        bool passed = !lowerAnswer.Contains(target, StringComparison.Ordinal);
        return new AssertionResult
        {
            Assertion = assertion,
            Passed = passed,
            Detail = passed
                ? $"✅ Correctly absent: \"{value}\""
                : $"❌ Unexpectedly found \"{value}\"",
        };
    }

    /// <summary>
    /// Evaluate every assertion attached to a row. Returns an empty list
    /// if the row has no assertions, or if the row's actual answer is
    /// empty or carries an <c>[ERROR:</c> prefix.
    /// </summary>
    public static List<AssertionResult> EvaluateRowAssertions(EvalRow row)
    {
        ArgumentNullException.ThrowIfNull(row);
        if (row.Assertions is null || row.Assertions.Count == 0)
        {
            return new List<AssertionResult>();
        }
        if (string.IsNullOrEmpty(row.ActualAnswer) || row.ActualAnswer.StartsWith("[ERROR:", StringComparison.Ordinal))
        {
            return new List<AssertionResult>();
        }
        var results = new List<AssertionResult>(row.Assertions.Count);
        foreach (Assertion a in row.Assertions)
        {
            results.Add(EvaluateAssertion(a, row.ActualAnswer));
        }
        return results;
    }

    /// <summary>
    /// Mutate every row: attach its assertion results in place. Returns
    /// the same list for fluent chaining.
    /// </summary>
    public static IList<EvalRow> EvaluateAllAssertions(IList<EvalRow> rows)
    {
        ArgumentNullException.ThrowIfNull(rows);
        foreach (EvalRow row in rows)
        {
            row.AssertionResults = EvaluateRowAssertions(row);
        }
        return rows;
    }
}
